.github/actions/build-test/entrypoint.sh

```diff
@@ -2,14 +2,15 @@
 
 cd /github/workspace/source && mkdir build && cd build || exit 1
 
-echo "Configuring cmake"
+echo ">>> Configuring cmake"
 cmake -DBUILD_CONTROLLERS=ON -DBUILD_DYNAMICAL_SYSTEMS=ON -DBUILD_ROBOT_MODEL=ON -DBUILD_TESTING=ON ..
 
-echo "Building project"
-make -j all
+echo ">>> Building project..."
+make -j all || (echo ">>> Build stage failed!" && exit 2) || exit $?
```

Comment:
That is interesting. I was not aware of this `||` command that executes only the second part if first one fails. But then I am not sure I understand the second block `|| exit $?`. Shouldn't it be executed only if `(echo ">>> Build stage failed!" && exit 2)` fails? But how come this actually fail?

Reply:
@domire8 here is the explanation :D

Reply:
I played around with the syntax for a while and ended up with this as the neatest working solution. 

As I understand it, there are two behaviours of `make` that need to be considered here. One is that make commands are not always fatal, so it doesn't automatically kill the entrypoint script. The second related behaviour is that `make` executes each section of the makefile in a subshell; the consequence of that is there's no guarantee that the last returned exit code (`$?`) refers to the part that failed non-fatally.

So, `make -j all || exit $?` does not always work. Similarly, neither does something like `if make -j all; then \n exit 1`. Somehow the double pipe OR `||` is able to catch the non-zero exit code properly. I imagine the `make` process has sub shells returning a number of exit codes, for example [0, 0, 0, 8, 0]. Then the OR of that output is "true", but the `if` and `$?` operations on the same output only see the last one, which is false (zero).

That's my theory at least from the reading and tests I did.

> Shouldn't it be executed only if (echo ">>> Build stage failed!" && exit 2) fails?
Yes, and it does fail because that sub shell has an `exit`, so it will trigger the subsequent `|| exit $?`. Here I'm using the last return code to avoid re-typing my custom error code (2 = build stage failure).

Related note, I could have just done `make -j all || exit 2`, but I added the sub shell with the echo so that it's easier to quickly search the output log for a message `>>> ...`. I cannot simply do `make -j all || echo ">>> Build stage failed!" && exit 2` since the order of operations means that, even on a successful make, it will always do the exit 2 at the end.

Reply:
I see thanks for the explanation. That is actually clever